### PR TITLE
Active Record hw

### PR DESCRIPTION
**Pull request by 3apy6a**
this is not the end result. I will add changes tomorrow

### CODE REVIEW

File: Active_Record/.env [hunk 1]
```diff
@@ -0,0 +1 @@
+RUBY_ENV=development
```
Comment (YuriiRudenko) [inline Active_Record/.env:1]: It's fine that this file is in your repository when it's a hometask. But in the real life DO NOT store this file in a repo

File: Active_Record/.idea/.rakeTasks [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (YuriiRudenko) [inline Active_Record/.idea/.rakeTasks]: remove folder `.idea` from a repo and add this folder to a `.gitignore`

File: Active_Record/Rakefile [hunk 3]
```diff
@@ -0,0 +1,56 @@
+require './config/initializers/database'
+
+namespace :db do
+
+  desc "Create the database"
+  task :create do
+    ActiveRecord::Base.establish_connection(CONFIG_DATABASE.except('database'))
+    ActiveRecord::Base.connection.create_database(CONFIG_DATABASE["database"])
```
Comment (YuriiRudenko) [inline Active_Record/Rakefile]: use single paws for a non-mutable strings

File: Active_Record/Rakefile [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    puts "Database created."
```
Comment (YuriiRudenko) [inline Active_Record/Rakefile]: same with paws

File: Active_Record/Rakefile [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  desc "Migrate the database"
```
Comment (YuriiRudenko) [inline Active_Record/Rakefile]: same with paws

File: Active_Record/Rakefile [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  task :migrate do
+    ActiveRecord::Tasks::DatabaseTasks.migrate
+    Rake::Task["db:schema"].invoke
+    puts "Database migrated."
+  end
+
+  desc 'Create a db/schema.rb file that is portable against any DB supported by AR'
+  task :schema do
+    require 'active_record/schema_dumper'
+    filename = "db/schema.rb"
```
Comment (YuriiRudenko) [inline Active_Record/Rakefile]: same with paws

File: Active_Record/Rakefile [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  task :create do
+    ActiveRecord::Base.establish_connection(CONFIG_DATABASE.except('database'))
+    ActiveRecord::Base.connection.create_database(CONFIG_DATABASE["database"])
+    puts "Database created."
+  end
+
+  desc "Migrate the database"
+  task :migrate do
+    ActiveRecord::Tasks::DatabaseTasks.migrate
+    Rake::Task["db:schema"].invoke
```
Comment (YuriiRudenko) [inline Active_Record/Rakefile]: same with paws

File: Active_Record/Rakefile [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    puts "Database migrated."
+  end
+
+  desc 'Create a db/schema.rb file that is portable against any DB supported by AR'
+  task :schema do
+    require 'active_record/schema_dumper'
+    filename = "db/schema.rb"
+    File.open(filename, "w:utf-8") do |file|
```
Comment (YuriiRudenko) [inline Active_Record/Rakefile]: same with paws

File: Active_Record/Rakefile [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+      ActiveRecord::SchemaDumper.dump(ActiveRecord::Base.connection, file)
+    end
+  end
+
+end
+
+namespace :g do
+  desc "Generate migration"
```
Comment (YuriiRudenko) [inline Active_Record/Rakefile]: same with paws

File: Active_Record/Rakefile [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  task :migration do
+    name = ARGV[1] || raise("Specify name: rake g:migration your_migration")
```
Comment (YuriiRudenko) [inline Active_Record/Rakefile]: same with paws

File: Active_Record/Rakefile [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    timestamp = Time.now.strftime("%Y%m%d%H%M%S")
```
Comment (YuriiRudenko) [inline Active_Record/Rakefile]: same with paws

File: Active_Record/app/models/booking.rb [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  validate :end_date_after_start_date
+
+  def booked_date_range
+    { from: :start_date, to: :end_date }
+  end
+
+  private
+
+  def end_date_after_start_date
+    return if end_date.blank? || start_date.blank?
```
Comment (YuriiRudenko) [inline Active_Record/app/models/booking.rb]: it is not necessary. This case is covered by `validates :start_date, :end_date, presence: true`

File: Active_Record/app/models/booking.rb [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+    if end_date < start_date
```
Comment (YuriiRudenko) [inline Active_Record/app/models/booking.rb]: use return
```ruby
return if end_date >= start_date
errors.add(:end_date, "must be after the start date")
```

File: Active_Record/app/models/booking.rb [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+      errors.add(:end_date, "must be after the start date")
```
Comment (YuriiRudenko) [inline Active_Record/app/models/booking.rb]: same with paws

File: Active_Record/app/models/booking_gears.rb [hunk 15]
```diff
@@ -0,0 +1,5 @@
+class BookingGear < ActiveRecord::Base
+  belongs_to :booking
+  belongs_to :gear
+
```
Comment (YuriiRudenko) [inline Active_Record/app/models/booking_gears.rb]: remove empty line

File: Active_Record/app/models/customer.rb [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+  has_many :bookings
+
+  validates :first_name, :last_name, presence: true
+  validates :desc, presence: true, length: { minimum: 10 }
+  validates :telephone_number, uniqueness: true
+
+  def full_name
+    "#{first_name} #{last_name}"
+  end
+
```
Comment (YuriiRudenko) [inline Active_Record/app/models/customer.rb:17]: remove empty line

File: Active_Record/app/models/gear.rb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+class Gear < ActiveRecord::Base
+  has_many :booking_gears, dependent: :destroy
+  has_many :bookings, through: :booking_gear
+
```
Comment (YuriiRudenko) [inline Active_Record/app/models/gear.rb:2]: remove empty line

File: Active_Record/app/models/booking_gears.rb [hunk 18]
```diff
@@ -0,0 +1,5 @@
+class BookingGear < ActiveRecord::Base
```
Comment (YuriiRudenko) [inline Active_Record/app/models/booking_gears.rb]: Convension over configuration. This model should be named `BookingsGear`

File: Active_Record/app/models/gear.rb [hunk 19]
```diff
@@ -0,0 +1,5 @@
+class Gear < ActiveRecord::Base
+  has_many :booking_gears, dependent: :destroy
+  has_many :bookings, through: :booking_gear
```
Comment (YuriiRudenko) [inline Active_Record/app/models/gear.rb]: you have no business logic in class `BookingGear`, so use `has_and_belongs_to_many`

File: Active_Record/app/models/sup.rb [hunk 20]
```diff
@@ -0,0 +1,10 @@
+class Sup < ActiveRecord::Base
+  has_many :customers, through: :bookings
+  has_many :bookings
```
Comment (YuriiRudenko) [inline Active_Record/app/models/sup.rb:4]: `has_many through` should be AFTER `has_many`, which declares   joiining relation

File: Active_Record/app/models/sup.rb [hunk 21]
```diff
@@ -0,0 +1,10 @@
...
+
+
```
Comment (YuriiRudenko) [inline Active_Record/app/models/sup.rb:8]: remove empty line

File: Active_Record/app/models/sup.rb [hunk 22]
```diff
@@ -0,0 +1,10 @@
...
+  def full_model
+    "SUP #{model} #{user_quantity}"
+  end
+
```
Comment (YuriiRudenko) [inline Active_Record/app/models/sup.rb]: remove empty line

File: Active_Record/config/initializers/initializers.rb [hunk 23]
```diff
@@ -0,0 +1,7 @@
+require './config/initializers/database'
+require './app/models/customer'
```
Comment (YuriiRudenko) [inline Active_Record/config/initializers/initializers.rb]: require dynamically

`Dir['./app/models/*.rb'].each { |f| require f.gsub('.rb', '') }`

File: Active_Record/db/migrate/20191208094805_create_sups.rb [hunk 24]
```diff
@@ -0,0 +1,17 @@
+class CreateSups < ActiveRecord::Migration[5.1]
+  
+  def up
+    create_table :sups do |t|
+      t.string :model
```
Comment (YuriiRudenko) [inline Active_Record/db/migrate/20191208094805_create_sups.rb]: It's a good practice if varchar fields have option `null: false, default: ''`

File: Active_Record/db/migrate/20191208094805_create_sups.rb [hunk 25]
```diff
@@ -0,0 +1,17 @@
+class CreateSups < ActiveRecord::Migration[5.1]
+
```
Comment (YuriiRudenko) [inline Active_Record/db/migrate/20191208094805_create_sups.rb]: remove empty line

File: Active_Record/db/migrate/20191208095100_create_customers.rb [hunk 26]
```diff
@@ -0,0 +1,16 @@
+class CreateCustomers < ActiveRecord::Migration[5.1]
+
```
Comment (YuriiRudenko) [inline Active_Record/db/migrate/20191208095100_create_customers.rb]: remove empty line

File: Active_Record/db/migrate/20191208095204_create_bookings.rb [hunk 27]
```diff
@@ -0,0 +1,17 @@
+class CreateBookings < ActiveRecord::Migration[5.1]
+
```
Comment (YuriiRudenko) [inline Active_Record/db/migrate/20191208095204_create_bookings.rb]: remove empty line

File: Active_Record/db/migrate/20191208095340_create_gears.rb [hunk 28]
```diff
@@ -0,0 +1,16 @@
+class CreateGears < ActiveRecord::Migration[5.1]
+
```
Comment (YuriiRudenko) [inline Active_Record/db/migrate/20191208095340_create_gears.rb]: remove empty line

File: Active_Record/db/migrate/20191208095340_create_gears.rb [hunk 29]
```diff
@@ -0,0 +1,16 @@
+class CreateGears < ActiveRecord::Migration[5.1]
+  
+  def up
+    create_table :gears do |t|
+      t.string :name
+      t.string :desc
+
+      t.timestamps
+    end
+
```
Comment (YuriiRudenko) [inline Active_Record/db/migrate/20191208095340_create_gears.rb]: remove empty line

File: Active_Record/db/migrate/20191208095417_create_events.rb [hunk 30]
```diff
@@ -0,0 +1,17 @@
+class CreateEvents < ActiveRecord::Migration[5.1]
+
```
Comment (YuriiRudenko) [inline Active_Record/db/migrate/20191208095417_create_events.rb]: remove empty line

File: Active_Record/db/migrate/20191208200645_create_booking_gears.rb [hunk 31]
```diff
@@ -0,0 +1,12 @@
+class CreateBookingGears < ActiveRecord::Migration[5.1]
+
```
Comment (YuriiRudenko) [inline Active_Record/db/migrate/20191208200645_create_booking_gears.rb]: remove empty line

File: Active_Record/db/migrate/20191208200645_create_booking_gears.rb [hunk 32]
```diff
@@ -0,0 +1,12 @@
...
+  
+  def up
+    create_table :booking_gears do |t|
+      t.references :booking, foreign_key: true
+      t.references :gear, foreign_key: true
+
+      t.timestamps
+    end
+  end
+
```
Comment (YuriiRudenko) [inline Active_Record/db/migrate/20191208200645_create_booking_gears.rb]: remove empty line

File: Active_Record/run.rb [hunk 33]
```diff
@@ -0,0 +1,18 @@
+require './config/initializers/initializers'
+
```
Comment (YuriiRudenko) [inline Active_Record/run.rb]: remove empty line

File: Active_Record/run.rb [hunk 34]
```diff
@@ -0,0 +1,18 @@
...
+
+sup_1 = Sup.new(
+    model: "Gladiator 10,6",
```
Comment (YuriiRudenko) [inline Active_Record/run.rb]: same with paws

File: Active_Record/run.rb [hunk 35]
```diff
@@ -0,0 +1,18 @@
...
+    avatar_url: "https://sup_photo.com",
```
Comment (YuriiRudenko) [inline Active_Record/run.rb]: same with paws

File: Active_Record/run.rb [hunk 36]
```diff
@@ -0,0 +1,18 @@
...
+    user_quantity: 1,
+    desc: "Oneuser SUP lenth -10,6"
+)
+puts "Sup: #{sup_1.id}"
+
+if sup_1.save
```
Comment (YuriiRudenko) [inline Active_Record/run.rb]: use short form
```ruby
puts sup_1.save ? 'Saved' : 'something went wrong'
```

File: Active_Record/run.rb [hunk 37]
```diff
@@ -0,0 +1,18 @@
...
+  puts 'Saved'
+else
+  puts "something went wrong"
```
Comment (YuriiRudenko) [inline Active_Record/run.rb]: same with paws

**Review by YuriiRudenko — CHANGES_REQUESTED** [review]

File: Active_Record/.idea/.rakeTasks (hunk 2 above)
Comment (3apy6a) [reply Active_Record/.idea/.rakeTasks]: done

File: Active_Record/Rakefile (hunk 3 above)
Comment (3apy6a) [reply Active_Record/Rakefile]: done

File: Active_Record/Rakefile (hunk 4 above)
Comment (3apy6a) [reply Active_Record/Rakefile]: done

File: Active_Record/Rakefile (hunk 5 above)
Comment (3apy6a) [reply Active_Record/Rakefile]: done

File: Active_Record/Rakefile (hunk 6 above)
Comment (3apy6a) [reply Active_Record/Rakefile]: done

File: Active_Record/Rakefile (hunk 7 above)
Comment (3apy6a) [reply Active_Record/Rakefile]: done

File: Active_Record/Rakefile (hunk 8 above)
Comment (3apy6a) [reply Active_Record/Rakefile]: done

File: Active_Record/Rakefile (hunk 9 above)
Comment (3apy6a) [reply Active_Record/Rakefile]: done

File: Active_Record/Rakefile (hunk 10 above)
Comment (3apy6a) [reply Active_Record/Rakefile]: done

File: Active_Record/Rakefile (hunk 11 above)
Comment (3apy6a) [reply Active_Record/Rakefile]: done

File: Active_Record/app/models/booking.rb (hunk 12 above)
Comment (3apy6a) [reply Active_Record/app/models/booking.rb]: done

File: Active_Record/app/models/booking.rb (hunk 14 above)
Comment (3apy6a) [reply Active_Record/app/models/booking.rb]: done

File: Active_Record/app/models/customer.rb (hunk 16 above)
Comment (3apy6a) [reply Active_Record/app/models/customer.rb:17]: done

File: Active_Record/app/models/gear.rb (hunk 17 above)
Comment (3apy6a) [reply Active_Record/app/models/gear.rb:2]: done

File: Active_Record/app/models/booking_gears.rb (hunk 18 above)
Comment (3apy6a) [reply Active_Record/app/models/booking_gears.rb]: does migration also get the wrong name?

File: Active_Record/app/models/sup.rb (hunk 20 above)
Comment (3apy6a) [reply Active_Record/app/models/sup.rb:4]: done

File: Active_Record/app/models/gear.rb (hunk 19 above)
Comment (3apy6a) [reply Active_Record/app/models/gear.rb]: done

File: Active_Record/app/models/sup.rb (hunk 21 above)
Comment (3apy6a) [reply Active_Record/app/models/sup.rb:8]: done

File: Active_Record/app/models/sup.rb (hunk 22 above)
Comment (3apy6a) [reply Active_Record/app/models/sup.rb]: done

File: Active_Record/app/models/booking.rb (hunk 13 above)
Comment (3apy6a) [reply Active_Record/app/models/booking.rb]: done

File: Active_Record/config/initializers/initializers.rb (hunk 23 above)
Comment (3apy6a) [reply Active_Record/config/initializers/initializers.rb]: done

File: Active_Record/run.rb (hunk 33 above)
Comment (3apy6a) [reply Active_Record/run.rb]: done

File: Active_Record/run.rb (hunk 34 above)
Comment (3apy6a) [reply Active_Record/run.rb]: done

File: Active_Record/run.rb (hunk 35 above)
Comment (3apy6a) [reply Active_Record/run.rb]: done

File: Active_Record/run.rb (hunk 36 above)
Comment (3apy6a) [reply Active_Record/run.rb]: ok, done

File: Active_Record/run.rb (hunk 37 above)
Comment (3apy6a) [reply Active_Record/run.rb]: done

File: Active_Record/app/models/booking_gears.rb (hunk 15 above)
Comment (3apy6a) [reply Active_Record/app/models/booking_gears.rb]: done

File: Active_Record/db/migrate/20191208094805_create_sups.rb (hunk 24 above)
Comment (3apy6a) [reply Active_Record/db/migrate/20191208094805_create_sups.rb]: done

File: Active_Record/db/migrate/20191208094805_create_sups.rb (hunk 25 above)
Comment (3apy6a) [reply Active_Record/db/migrate/20191208094805_create_sups.rb]: done

File: Active_Record/db/migrate/20191208095100_create_customers.rb (hunk 26 above)
Comment (3apy6a) [reply Active_Record/db/migrate/20191208095100_create_customers.rb]: done

File: Active_Record/db/migrate/20191208095204_create_bookings.rb (hunk 27 above)
Comment (3apy6a) [reply Active_Record/db/migrate/20191208095204_create_bookings.rb]: done

File: Active_Record/db/migrate/20191208095340_create_gears.rb (hunk 28 above)
Comment (3apy6a) [reply Active_Record/db/migrate/20191208095340_create_gears.rb]: done

File: Active_Record/db/migrate/20191208095340_create_gears.rb (hunk 29 above)
Comment (3apy6a) [reply Active_Record/db/migrate/20191208095340_create_gears.rb]: done

File: Active_Record/db/migrate/20191208095417_create_events.rb (hunk 30 above)
Comment (3apy6a) [reply Active_Record/db/migrate/20191208095417_create_events.rb]: done

File: Active_Record/db/migrate/20191208200645_create_booking_gears.rb (hunk 31 above)
Comment (3apy6a) [reply Active_Record/db/migrate/20191208200645_create_booking_gears.rb]: done

File: Active_Record/db/migrate/20191208200645_create_booking_gears.rb (hunk 32 above)
Comment (3apy6a) [reply Active_Record/db/migrate/20191208200645_create_booking_gears.rb]: done

File: Active_Record/app/models/booking_gears.rb (hunk 18 above)
Comment (3apy6a) [reply Active_Record/app/models/booking_gears.rb]: done

File: Active_Record/app/models/customer.rb [hunk 38]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+class Customer < ActiveRecord::Base
+  has_many :bookings
+  has_many :sups, through: :bookings
+
+  before_validation :normalize_name, on: :create
+
+  validates :first_name, :last_name, :telephone_number, presence: true
+  validates :description, presence: true, length: {minimum: 10 }
```
Comment (YuriiRudenko) [inline Active_Record/app/models/customer.rb:10]: space after brace

File: Active_Record/run.rb [hunk 39]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+require './config/initializers/initializer'
+sup_1 = Sup.new(
+        model: 'Red Paddle Voyager 13,2"',
```
Comment (YuriiRudenko) [inline Active_Record/run.rb:4]: spaces

**Pull request merged**